log4j-core-test/src/test/java/org/apache/logging/log4j/core/appender/SocketAppenderReconnectTest.java

```diff
@@ -0,0 +1,387 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.core.appender;
+
+import static org.apache.logging.log4j.core.appender.SslContexts.createSslContext;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.net.ssl.SSLContext;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.LoggerContext;
+import org.apache.logging.log4j.core.config.Configuration;
+import org.apache.logging.log4j.core.config.Configurator;
+import org.apache.logging.log4j.core.config.builder.api.AppenderComponentBuilder;
+import org.apache.logging.log4j.core.config.builder.api.ComponentBuilder;
+import org.apache.logging.log4j.core.config.builder.api.ConfigurationBuilder;
+import org.apache.logging.log4j.core.config.builder.api.ConfigurationBuilderFactory;
+import org.apache.logging.log4j.core.config.builder.impl.BuiltConfiguration;
+import org.apache.logging.log4j.core.net.TcpSocketManager;
+import org.apache.logging.log4j.core.net.TcpSocketManager.HostResolver;
+import org.apache.logging.log4j.core.net.ssl.SslKeyStoreConstants;
+import org.apache.logging.log4j.test.junit.UsingStatusListener;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.CleanupMode;
+import org.junit.jupiter.api.io.TempDir;
+
+/**
+ * Tests reconnection support of {@link org.apache.logging.log4j.core.appender.SocketAppender}.
+ */
+public class SocketAppenderReconnectTest {
+
+    private static final int EPHEMERAL_PORT = 0;
+
+    private static final String APPENDER_NAME = "TestSocket";
+
+    /**
+     * Tests if failures are propagated when reconnection fails.
+     *
+     * @see <a href="https://issues.apache.org/jira/browse/LOG4J2-2829">LOG4J2-2829</a>
+     */
+    @Test
+    @UsingStatusListener // Suppresses `StatusLogger` output, unless there is a failure
+    void repeating_reconnect_failures_should_be_propagated() throws Exception {
+        try (final LineReadingTcpServer server = new LineReadingTcpServer()) {
+
+            // Start the server.
+            server.start("Main", EPHEMERAL_PORT);
+            final int port = server.getServerSocket().getLocalPort();
+
+            // Initialize the logger context
+            final Configuration config = createConfiguration(port, null);
+            try (final LoggerContext loggerContext = Configurator.initialize(config)) {
```

Comment:
> Since you don't call `LogManager`, why don't you create a `LoggerContext` that is not provided by the context selector?

You're right. Fixed in 68e8c230eb184b49d719b6b450f5da878f2301fd.

> You could even use `@LoggerContextSource` to create such a context.

I will address this in [the programmatic configuration conversation](https://github.com/apache/logging-log4j2/pull/2767#discussion_r1744973471).